### PR TITLE
CLI refactor with commander

### PR DESCRIPTION
**Pull request by pitaj**
This is a large diff. Most of the new files are composed of code moved out of `nodebb` or `app` with small changes here and there.

- Modularized the functionality
- All functionality done directly from `./nodebb` now
(still available from `app` for backwards compatibility)
- Moved all CLI code from `./nodebb` to `src/cli`
- Fixed `nodebb.bat` to work from any location, like `./nodebb`, and
also hides command output
- Overwrite some commander methods to add CLI color support
- Added `./nodebb info` for quick info including git hash, NodeBB
version, node version, and some database info
- Refactored `./nodebb reset` to allow multiple resets at once
- Changed `./nodebb restart` to essentially stop and start, as Windows
doesn't support signals
- Added `-l, --log` option which works on `./nodebb start` and `./nodebb
restart` to show logging, like `./nodebb slog`
- Expanded `-d, --dev` option which works on them as well, like
`./nodebb dev`
- Improvements to self-help. `./nodebb build -h` will output all
possible targets
- `./nodebb reset` explains usage better

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -89,224 +59,40 @@ if (!process.send) {
...
-	async.waterfall([
-		async.apply(require('./src/reset').reset),
-		async.apply(require('./src/meta/build').buildAll),
+	var options = {
+		theme: nconf.get('t'),
+		plugin: nconf.get('p'),
+		widgets: nconf.get('w'),
+		settings: nconf.get('s'),
+		all: nconf.get('a'),
+	};
```
Comment (julianlam) [inline app.js]: :clap:

File: app.js [hunk 2]
```diff
@@ -89,224 +59,40 @@ if (!process.send) {
...
+
+	async.series([
+		async.apply(require('./src/cli/reset').reset, options),
+		require('./src/meta/build').buildAll,
 	], function (err) {
-		process.exit(err ? 1 : 0);
+		if (err) {
+			throw err;
```
Comment (julianlam) [inline app.js]: Is outputting a stack trace the best course of action here? Wouldn't a graceful error handle work better?

I say this knowing full well that the old code didn't handle it any better :smile:

File: app.js [hunk 3]
```diff
@@ -89,224 +59,40 @@ if (!process.send) {
...
-	var version = process.version.slice(1);
-	var range = pkg.engines.node;
-	var semver = require('semver');
-	var compatible = semver.satisfies(version, range);
-
-	if (!compatible) {
-		winston.warn('Your version of Node.js is too outdated for NodeBB. Please update your version of Node.js.');
-		winston.warn('Recommended ' + range.green + ', '.reset + version.yellow + ' provided\n'.reset);
-	}
-}
```
Comment (julianlam) [inline app.js]: :+1:

File: src/cli/index.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+program
+	.name('./nodebb')
+	.description('Welcome to NodeBB')
+	.version(pkg.version)
+	.option('--json-logging', 'Output to logs in JSON format', false)
+	.option('--log-level <level>', 'Default logging level to use', 'info')
+	.option('-d, --dev', 'Development mode, including verbose logging', false)
+	.option('-l, --log', 'Log subprocess output to console', false)
+	.option('-c, --config <value>', 'Specify a config file', 'config.json')
```
Comment (julianlam) [inline src/cli/index.js]: `--no-silent` and `--no-daemon`?

**Comment by julianlam** [comment]
Multiple plugin resets via shorthand fail:

```
julian@arya ~/Projects/nodebb/forum (commander) $ ./nodebb reset -p mentions,markdown
21/11 10:23:09 [18981] - warn: [reset] Plugin `nodebb-plugin-mentions,markdown` was not active on this forum
21/11 10:23:09 [18981] - info: [reset] No action taken.
21/11 10:23:09 [18981] - error: [reset] Errors were encountered during reset Error: plugin-not-active
    at /home/julian/Projects/nodebb/forum/src/cli/reset.js:151:10
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:421:16
    at next (/home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:5302:29)
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:906:16
    at /home/julian/Projects/nodebb/forum/src/events.js:36:3
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:421:16
    at next (/home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:5302:29)
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:906:16
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:3861:9
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:421:16
/home/julian/Projects/nodebb/forum/src/cli/reset.js:76
                        throw err;
                        ^

Error: plugin-not-active
    at /home/julian/Projects/nodebb/forum/src/cli/reset.js:151:10
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:421:16
    at next (/home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:5302:29)
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:906:16
    at /home/julian/Projects/nodebb/forum/src/events.js:36:3
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:421:16
    at next (/home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:5302:29)
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:906:16
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:3861:9
    at /home/julian/Projects/nodebb/forum/node_modules/async/dist/async.js:421:16
```

**Comment by julianlam** [comment]
`./nodebb help <cmd>` doesn't seem to work, outputs regular help screen.

```
$ ./nodebb help start

  Usage: ./nodebb [options] [command]

  Welcome to NodeBB


  Options:

    -V, --version         output the version number
    --json-logging        Output to logs in JSON format
    --log-level <level>   Default logging level to use
    -d, --dev             Development mode, including verbose logging
    -l, --log             Log subprocess output to console
    -c, --config <value>  Specify a config file
    -h, --help            output usage information


  Commands:

    start                 Start the NodeBB server
    stop                  Stop the NodeBB server
    restart               Restart the NodeBB server
    status                Check the running status of the NodeBB server
    log                   Open the output log (useful for debugging)
    setup                 Run the NodeBB setup script
    install               Launch the NodeBB web installer for configuration setup
    build [targets...]    Compile static assets (JS, CSS, templates, languages, sounds)
    activate [plugin]     Activate a plugin for the next startup of NodeBB (nodebb-plugin- prefix is optional)
    plugins               List all installed plugins
    events                Outputs the last ten (10) administrative events recorded by NodeBB
    info                  Outputs various system info
    reset [options]       Reset plugins, themes, settings, etc
    upgrade [scripts...]  Run NodeBB upgrade scripts and ensure packages are up-to-date, or run a particular upgrade script
    help [cmd]            display help for [cmd]
```

File: src/cli/index.js (hunk 4 above)
Comment (julianlam) [reply src/cli/index.js]: No that's fine, if they aren't passed through right now, no need to do it :smile:

File: app.js (hunk 2 above)
Comment (pitaj) [reply app.js]: I think outputting a stack trace is always more useful than _just_ outputting an error message. If something is wrong, we should detect it before an error is thrown in the first place.

**Comment by pitaj** [comment]
@julianlam the newest commit adds the ability to specify what steps of the upgrade should be run. To run only the package updates, npm installs, and plugin updates: `./nodebb upgrade -mip`

**Comment by julianlam** [comment]
Does this mean ./nodebb upgradePlugins no longer works? That's an acceptable compromise imo

**Comment by pitaj** [comment]
It still works, but `./nodebb upgrade -p` is preferable.

**Pull request merged**